src/components/sketches/Scene5_happy.ts

```diff
@@ -0,0 +1,290 @@
+      this.firework = firework;
+      this.lifespan = FIREWORK_LIFESPAN;
+      this.acc = p.createVector(0, 0);
+      this.ran = p.random(0, 3);
+      if (this.firework) {
+        this.vel = p.createVector(0, p.random(-12, -8));
+      } else {
+        this.vel = p5.Vector.random2D();
+        this.vel.mult(p.random(2, 10));
+      }
+    }
+
+    applyForce(force: p5.Vector) {
+      this.acc.add(force);
+    }
+
+    update() {
+      if (!this.firework) {
+        this.vel.mult(0.9);
+        this.lifespan -= 4;
+      }
+
+      this.vel.add(this.acc);
+      this.pos.add(this.vel);
+      this.acc.mult(0);
+    }
+
+    done() {
+      if (this.lifespan < 0) {
+        return true;
+      } else {
+        return false;
+      }
+    }
+
+    show() {
+      if (!this.firework) {
+        // after explosion
+        p.strokeWeight(6);
+        p.stroke(255, 211, 97, this.lifespan);
+      } else {
+        p.strokeWeight(8);
+      }
+
+      if (this.ran < 1) p.stroke('rgba(255,211,97,0.2)');
+      else if (this.ran < 2) p.stroke('rgba(255,211,97,0.4)');
+      else p.stroke('rgba(255,211,97,0.6)');
+
+      p.point(this.pos.x, this.pos.y);
+    }
+  }
+
+  class ringParticle {
+    pos: p5.Vector;
+    vel: p5.Vector;
+    acc: p5.Vector;
+    r: number;
+    time: number;
+    explode: boolean;
+    lifespan: number;
+    history: p5.Vector[];
+
+    constructor(x: number, y: number, rot: number) {
+      this.pos = p.createVector(x, y); // Position
+      this.vel = p.createVector(2, 2); // Velocity
+      this.r = RING_RADIUS; //3*noise(this.pos.x/250, this.pos.y/250)*5;
+      this.vel.limit(5);
+      this.acc = p.createVector(0.1, 0.1); // Acceleration
+      this.vel.rotate(rot); // Rotate velocity to the given rotation (avoids complicated maths)
+      this.acc.rotate(rot); // Rotate to the given rotation (avoids complicated maths)
+      this.time = 0; // Time alive
+      this.explode = false;
+
+      this.lifespan = FIREWORK_LIFESPAN;
+      this.history = [];
+    }
+    update() {
+      this.vel.add(this.acc);
+      this.pos.add(this.vel); // Update position with velocity
+      // If older than 40, be deleted next draw event
+      if (this.time <= 40) {
+        // If older than 15, reduce the size of the ringParticles every step
+        if (this.time > 15) {
+          this.r -= 0.05;
+        }
+        this.time++;
+      } else {
+        if (this.lifespan == FIREWORK_LIFESPAN) {
+          this.explode = true;
+          this.applyForce(p.createVector(0, 0.3));
+
+          this.vel = p5.Vector.random2D();
+          this.vel.mult(p.random(2, 10));
+        }
+        this.vel.mult(VELOCITY_CHANGE); // vel change ***
+        this.lifespan -= LIFESPAN_DECREMENT; // lifespan ***
+
+        const v = p.createVector(this.pos.x, this.pos.y);
+        this.history.push(v);
+
+        //size of trail ***
+        if (this.history.length > TRAIL_SIZE) {
+          this.history.splice(0, 1);
```

Comment:
try to see if you can use `Array.pop()` it's O(1)

Reply:
I can't here. Tried `this.history.pop()`, but it freezes the firework in time because pop deletes last element. I just want to delete first element because it is oldest in the history. If I draw and delete over and over again at the last element, there will be no drawings be placed (trails) after history reaches trail_size.

Reply:
I could do `shift()` , but that is O(n) as well

Reply:
Hmmm ideally we should create a queue that pushes in O(1) and pops in O(1)... we can do this optimization later with a linkedlist implementation of a queue but for now I'm good with this!

Can you make an issue for this optimization?

Reply:
yessir, yea a queue would be ideal.